### PR TITLE
Adding SurfaceButtonInteract Event inside Warhead

### PR DESCRIPTION
**Pull request by SlejmUr**
Adding more funcionality inside warhead, currently this will trigger after the distance check and before the keycard permission check.

### CODE REVIEW

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    [EventPatch(typeof(Warhead), nameof(Warhead.SurfaceButtonInteract))]
+    [HarmonyPatch(typeof(PlayerInteract), nameof(PlayerInteract.UserCode_CmdSwitchAWButton))]
+    internal static class SurfaceButtonInteract
+    {
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            // Getting the start of BypassMode (IL_0028: ldarg.0)
+            int index = newInstructions.FindLastIndex(instruction => instruction.ToString().Contains("BypassMode")) - 2;
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:37]: `int index = newInstructions.FindIndex(instruction => instruction.LoadsField(Field(typeof(AlphaWarheadOutsitePanel), nameof(AlphaWarheadOutsitePanel.keycardEntered)))) + 3;`

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                    new CodeInstruction(OpCodes.Dup),
+
+                    // Warhead.OnSurfaceButtonInteract(ev)
+                    new CodeInstruction(OpCodes.Call, Method(typeof(Warhead), nameof(Warhead.OnSurfaceButtonInteract))),
+                    new CodeInstruction(OpCodes.Callvirt, PropertyGetter(typeof(SurfaceButtonInteractEventArgs), nameof(SurfaceButtonInteractEventArgs.IsAllowed))),
+
+                    // if (!ev.IsAllowed)
+                    // return;
+                    new CodeInstruction(chckDis_instruction),
+                    new CodeInstruction(OpCodes.Ret),
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:74]: use `new(OpCodes.Brfalse_S, thisFunctionLabel)`

Comment (xNexusACS) [reply Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:74]: remove both instructions

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                });
+
+            // Getting the new index of our function.
+            int hub_indx = newInstructions.FindLastIndex(instruction => instruction.ToString().Contains("Player::Get")) - 2;
+            newInstructions[hub_indx].WithLabels(thisFunctionLabel);
+            chckDis_instruction.operand = thisFunctionLabel;
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:80]: this 3 lines can be removed

use instead
`newInstructions[newInstructions.Count - 1].labels.Add(thisFunctionLabel);`

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            // if (!ev.IsAllowed)
+            // return;
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    new CodeInstruction(OpCodes.Ldarg_0),
+
+                    // PlayerInteract._hub
+                    new CodeInstruction(OpCodes.Ldfld, Field(typeof(PlayerInteract), "_hub")),
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:56]: `new CodeInstruction(OpCodes.Ldfld, Field(typeof(PlayerInteract), nameof(PlayerInteract._hub))),`

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstruction> instructions, ILGenerator generator)
+        {
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            // Getting the start of BypassMode (IL_0028: ldarg.0)
+            int index = newInstructions.FindLastIndex(instruction => instruction.ToString().Contains("BypassMode")) - 2;
+
+            // Getting the ChckDis jump to the new label. (IL_0025: brtrue.s)
+            int chckDis_call_after_indx = newInstructions.FindLastIndex(instruction => instruction.ToString().Contains("ChckDis")) + 1;
+            CodeInstruction chckDis_instruction = newInstructions[chckDis_call_after_indx];
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:41]: both lines can be removed

File: Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+            // SurfaceButtonInteractEventArgs ev = new SurfaceButtonInteractEventArgs(Player.Get(PlayerInteract._hub), bool)
+            // Warhead.OnSurfaceButtonInteract(ev)
+            // if (!ev.IsAllowed)
+            // return;
+            newInstructions.InsertRange(
+                index,
+                new[]
+                {
+                    new CodeInstruction(OpCodes.Ldarg_0),
```
Comment (xNexusACS) [inline Exiled.Events/Patches/Events/Warhead/SurfaceButtonInteract.cs:53]: `new CodeInstruction(OpCodes.Ldarg_0).MoveLabelsFrom(newInstructions[index]),`

**Review by xNexusACS — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Warhead/SurfaceButtonInteractEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            Player = player;
+            IsAllowed = isAllowed;
+        }
+
+        /// <inheritdoc/>
+        public bool IsAllowed { get; set; }
+
+        /// <summary>
+        /// Gets the player who's interacting the warhead button in surface.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Warhead/SurfaceButtonInteractEventArgs.cs:37]:
```suggestion
        /// Gets the player who's interacting with the Warhead button in surface.
```

File: Exiled.Events/EventArgs/Warhead/SurfaceButtonInteractEventArgs.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+// </copyright>
+// -----------------------------------------------------------------------
+
+namespace Exiled.Events.EventArgs.Warhead
+{
+    using API.Features;
+    using Interfaces;
+
+    /// <summary>
+    /// Contains all information before a player interacts the Warhead button in surface.
```
Comment (NaoUnderscore) [inline Exiled.Events/EventArgs/Warhead/SurfaceButtonInteractEventArgs.cs:14]:
```suggestion
    /// Contains all information before a player interacts with the Warhead button in surface.
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Review by xNexusACS — APPROVED** [review]

**Pull request merged**